kimchi/src/folding/expressions.rs

```diff
@@ -0,0 +1,305 @@
+use super::{CacheId, GateType};
+use crate::circuits::{
+    expr::{Column, Expr, FeatureFlag, Op2, Variable},
+    gate::CurrOrNext,
+};
+
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub enum FoldingColumn {
+    Witness(usize),
+    ///for the extra columns added by quadricization
+    #[allow(dead_code)]
+    WitnessExtended(usize),
+    Index(GateType),
+    Coefficient(usize),
+    ///basically X, to allow accesing the next row
+    Shift,
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+///A simplified expression to be use in folding
+pub enum FoldingExp<C> {
+    Constant(C),
+    Cell(FoldingColumn),
+    Double(Box<FoldingExp<C>>),
+    Square(Box<FoldingExp<C>>),
+    Add(Box<FoldingExp<C>>, Box<FoldingExp<C>>),
+    Sub(Box<FoldingExp<C>>, Box<FoldingExp<C>>),
+    Mul(Box<FoldingExp<C>>, Box<FoldingExp<C>>),
+    UnnormalizedLagrangeBasis(i32),
+    Cache(CacheId, Box<FoldingExp<C>>),
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub enum Degree {
+    Zero,
+    One,
+    Two,
+}
+impl<C> FoldingExp<C> {
+    pub(super) fn degree(&self) -> Degree {
+        use Degree::*;
+        match self {
+            FoldingExp::Constant(_) => Zero,
+            FoldingExp::Cell(col) => match col {
+                FoldingColumn::Witness(_) | FoldingColumn::WitnessExtended(_) => One,
+                _ => Zero,
+            },
+            FoldingExp::Double(e) => e.degree(),
+            FoldingExp::Square(e) => &e.degree() * &e.degree(),
+            FoldingExp::Add(e1, e2) | FoldingExp::Sub(e1, e2) | FoldingExp::Mul(e1, e2) => {
+                e1.degree() + e2.degree()
+            }
+            FoldingExp::UnnormalizedLagrangeBasis(_) => Zero,
+            FoldingExp::Cache(_, e) => e.degree(),
+        }
+    }
+}
+impl std::ops::Add for Degree {
+    type Output = Self;
+
+    fn add(self, rhs: Self) -> Self::Output {
+        use Degree::*;
+        match (self, rhs) {
```

Comment:
The idea is to support only degree $<=2$ at folding level, and to support arbitrary degree by converting expressions above it into degree 2 by using quadricization.
But maybe it would be better to do the quadricization work internally inside folding, still this type in particular would only handle degree $<=2$ or maybe $<=3$ if I implement and optimization I thought for the challenges.

Reply:
Can you elaborate a bit more? I'm not sure how quadricization works, maybe there's somewhere I can read about it?

Reply:
the assumption would be for higher degree expressions to not appear, with the detail that the degree is computed only counting the witness columns, fixed columns are "free".
But I think I'll integrate quadricization into folding, that should make everything clearer.

Reply:
do you assert somewhere degree of input is $\le 2$